### ISSUE TITLE
scoped seems not work on complex graph?

### ISSUE THREAD

**Issue by duongphuhiep** [firasdarwish/ore#2]
My dependency graph is quite complex:

![complex dependency graph](https://github.com/user-attachments/assets/0e624097-9b86-487c-99c3-1f1c4db8041d)

In the same "context", I expected that ore will give me 2 different instances of "A" (because "A" is transient) and the same instance of "C" (because "C" is scoped). But I got 2 instances of C (C-10 and C-17 in the same scope):

```go
a1, ctx := ore.Get[*demo1.A](ctx)
log.Println(a1.ToString())  //got: A-11 { B-9 { D-3 { F-1, H-2 }, E-8{ Gb-4, Gc-5, DGa-7 { Ga-6 },  } }, C-10 }
a2, _ := ore.Get[*demo1.A](ctx) //got:  A-18 { B-16 { D-13 { F-12, H-2 }, E-15{ Gb-14, Gc-5, DGa-7 { Ga-6 },  } }, C-17 }
log.Println(a2.ToString())
```

[This is the full codes to reproduce the problem](https://github.com/duongphuhiep/go-learn/blob/ore_issue/try-uberfx/main.go#L13)

Note that: I got the expected behaviour when keeping only the transient "A" and the scoped "C" and remove everything else. [This test passed without problem](https://github.com/duongphuhiep/go-learn/blob/ore_issue/try-uberfx/demo1/OreScoped_test.go#L17)

The problem happened only on complex graph. Am I missing something? or is it a bug?

**Comment by firasdarwish** [comment]
hi @duongphuhiep 
first of all, thank you for reporting

as i noticed you're expecting `a1` and `a2` to be the same instance. this can be done by registering `*demo1.A` as `Scoped` and then reusing the same `context` (which is done correctly)

but the issue is that you're registering `*demo1.A` as `Transient` in `try-uberfx/demo1/DiOre.go`

```go
func RegisterDependenciesToOre_UseFunc() {
	ore.RegisterLazyFunc(ore.Transient, func(ctx context.Context) *A {
		b, ctx := ore.Get[*B](ctx)
		c, _ := ore.Get[*C](ctx)
		return NewA(b, c)
	})
```

please try changing it to `Scoped`:
```go
func RegisterDependenciesToOre_UseFunc() {
	ore.RegisterLazyFunc(ore.Scoped, func(ctx context.Context) *A {
		b, ctx := ore.Get[*B](ctx)
		c, _ := ore.Get[*C](ctx)
		return NewA(b, c)
	})
```
-------------------------

i noticed you're doing some tests and comparisons of multiple DI solutions for Go, i'd be glad to hear your review and suggestions for Ore, whether fixing some bugs or adding new features

**Comment by firasdarwish** [comment]
hi @duongphuhiep 
first of all, thank you for reporting

there was a bug that prevented reusing the `context` when resolving nested services
a fix has been merged and released with v0.2.0
and i edited your code example accordingly and submitted a [PR](https://github.com/duongphuhiep/go-learn/pull/1) 

i noticed you're doing some tests and comparisons of multiple DI solutions for Go, i'd be glad to hear your review and suggestions for Ore, whether fixing some bugs or adding new features

**Comment by duongphuhiep** [comment]
thanks for your answer and for the hard work on this fantastic library. I will test some more and link my review to you when finished.

As feature requests. I wish that Ore can do something like this

### ore.Validate

Run on application startup to detect problems on the DI graph before it happens on runtime

```go
ore.Validate(ValidateOptions{checkDepsLifeTime: true, checkCircularDeps: true, checkMissingDeps: true})
```

* check if a service with longer lifetime should not depend on a services with shorter lifetime. For eg a Singleton depends on Scoped or Transient is a big security risk.
* warning about circular deps
* warning about missing deps

### Service aliasing

1) Implicite: Register struct, Get interfaces

```go
//MyDisposableEngine implements IDisposable and IEngine
ore.RegisterLazyFunc(ore.Singleton, func(ctx context.Context) (*MyDisposableEngine, context.Context) {
  return MyDisposableEngine(), ctx
})
disposable, _ := ore.Get[IDisposable](ctx)
engine, _ := ore.Get[IEngine](ctx)
```

2) Or Explicit

```go
//MyDisposableEngine implements IDisposable and IEngine
ore.RegisterLazyFunc(ore.Singleton, func(ctx context.Context) (*MyDisposableEngine, context.Context) {
  return MyDisposableEngine(), ctx
}).As[IDisposable]().As[IEngine]()
disposable, _ := ore.Get[IDisposable](ctx)
disposable, _ := ore.Get[IEngine](ctx)
```
